test/web-platform-tests/run-wpts.js

```diff
@@ -61,29 +63,48 @@ describe("web-platform-tests", () => {
           });
 
           const testFile = testFilePath.slice((toRunDoc.DIR + "/").length);
-          let reason;
+          let reason, data;
 
           if (matchingPattern) {
+            // The array case is when the failure affects the whole test file
+            // (ex.: testharness timeout or error, uncaught exception, etc.)
             reason = toRunDoc[matchingPattern][0];
+            if (!Array.isArray(toRunDoc[matchingPattern])) {
+              // The non-array case is when some subtests in the test file pass,
+              // but others fail, and testharness status is OK.
+              data = toRunDoc[matchingPattern];
+            }
           } else if (toRunDoc.DIR.startsWith("html/canvas/")) {
             reason = "needs-canvas";
           }
 
-          const shouldSkip = ["fail-slow", "timeout", "flaky"].includes(reason) ||
-                             (["fail-with-canvas", "needs-canvas"].includes(reason) && !hasCanvas);
-          const expectFail = (reason === "fail") ||
-                             (reason === "fail-with-canvas" && hasCanvas) ||
-                             (reason === "needs-node10" && !hasNode10) ||
-                             (reason === "needs-node11" && !hasNode11) ||
-                             (reason === "needs-node12" && !hasNode12);
-
-          if (shouldSkip) {
-            specify.skip(`[${reason}] ${testFile}`);
-          } else if (expectFail) {
-            const failReason = reason !== "fail" ? `: ${reason}` : "";
-            runSingleWPT(testFilePath, `[expected fail${failReason}] ${testFile}`, expectFail);
-          } else {
-            runSingleWPT(testFilePath, testFile, expectFail);
+          switch (resolveReason(reason)) {
+            case "skip": {
+              specify.skip(`[${reason}] ${testFile}`);
+              break;
+            }
+
+            case "expect-fail": {
+              const failReason = reason !== "fail" ? `: ${reason}` : "";
+              runSingleWPT(testFilePath, `[expected fail${failReason}] ${testFile}`, true);
+              break;
+            }
+
+            default: {
+              let failCount = 0;
+              if (data) {
+                failCount = Object.values(data).reduce((count, innerReason) => {
```

Comment:
How about `Object.values(data).filter(innerReason => resolveReason(innerReason) === "expect-fail").length` ?

Reply:
I guess that that reads nicer, but it requires allocating an extra intermediate array, which is one of the main reason I’m looking forward to when [the **Iterator Helpers** proposal](https://github.com/tc39/proposal-iterator-helpers) reaches stage 4 and gets implemented in all engines.